src/twisted/mail/bounce.py

```diff
@@ -7,64 +7,74 @@
 """
 Support for bounce message generation.
 """
-import StringIO
 import email.utils
 import time
 import os
-
+from io import StringIO, SEEK_SET, SEEK_END
 
 from twisted.mail import smtp
 
-BOUNCE_FORMAT = """\
-From: postmaster@%(failedDomain)s
-To: %(failedFrom)s
+BOUNCE_FORMAT = u"""\
+From: postmaster@{failedDomain}
+To: {failedFrom}
 Subject: Returned Mail: see transcript for details
-Message-ID: %(messageID)s
+Message-ID: {messageID}
 Content-Type: multipart/report; report-type=delivery-status;
-    boundary="%(boundary)s"
+    boundary="{boundary}"
 
---%(boundary)s
+--{boundary}
 
-%(transcript)s
+{transcript}
 
---%(boundary)s
+--{boundary}
 Content-Type: message/delivery-status
-Arrival-Date: %(ctime)s
-Final-Recipient: RFC822; %(failedTo)s
+Arrival-Date: {ctime}
+Final-Recipient: RFC822; {failedTo}
 """
 
 
 
-def generateBounce(message, failedFrom, failedTo, transcript=''):
+def generateBounce(message, failedFrom, failedTo, transcript='',
+                   encoding='utf-8'):
     """
     Generate a bounce message for an undeliverable email message.
 
-    @type message: L{bytes}
+    @type message: a file-like object
```

Comment:
Isn't this change backwards incompatible?

Reply:
In addition to porting, I added more tests to increase the code coverage.  As I increased the code coverage, I found a few small bugs.  For example, `message` could never have been bytes.  If you look at this line:
https://github.com/twisted/twisted/blob/trunk/src/twisted/mail/bounce.py#L81

The code is performing `seek()` and `tell()` on `message`, so `message` could never have been bytes.  It had to have been a file like object.  In this change I am just clarifying the doc string to match the reality of what this code has been doing for a long time.